### ISSUE TITLE
TripleOscillator: Oscillators are getting out of sync

### ISSUE THREAD

**Issue by musikBear** [LMMS/lmms#2047]
I have spend some time trying to figure out, why lmms sounds less 'punchy/ agressive' than other daw's
during that i have seen something strange
To reproduce
insert a std 3oc
set all 3 osc to saw
insert a 18 bar long note in D3
play
The sound variates in a sweep-like fashion:
![paste](https://cloud.githubusercontent.com/assets/6662066/7587156/79f8c944-f8b4-11e4-8856-e39e8fb08444.PNG)

Why?
There are no 
- envelope settings
- filters
- reso

why is this note not completely uniform in spektrum?
If i do exactly the same with 3oc in fls-demo, i get a completely uniform spectrum

![paste](https://cloud.githubusercontent.com/assets/6662066/7587173/a4c4afee-f8b4-11e4-8bfa-75d414a263d3.png)

**Comment by Sti2nd** [comment]
> why lmms sounds less 'punchy/ agressive' than other daw's

Maybe you haven't removed the default limiter from FL? I haven't got a clue how synthesis really work, but I can easily imagine that there are different algorithms for making a sine wave... TripleOscillator for LMMS is not the same instrument as 3OSC for FL!

**Comment by michaelgregorius** [comment]
@Sti2nd I think it's really a bug (if it is not an intended feature).

To get an idea of what might be causing this behaviour and to simplify the conditions I have set all three oscillators to play a sine wave instead of a saw. The same effect can still be observed which means that we can likely eliminate the aliasing in the saw oscillators as a possible cause. My guess is that the three oscillators go out of phase over time and that the observed lowering of amplitude is the effect of some destructive interference due to the phasing.

Here is the sum of all three sines (fundamental + octave + 2. octave) at the beginning of the note (also the first samples of the test render):
![2047-start](https://cloud.githubusercontent.com/assets/9293269/7595596/41cc95ee-f8e8-11e4-95ff-8736135a6c35.png)

Please note that the waveform is symmetric at this point in time and that it looks like expected. And now here is the waveform towards the end of my rendered wav file:
![2047-end](https://cloud.githubusercontent.com/assets/9293269/7595612/5d37a300-f8e8-11e4-8e86-79007c38075f.png)

Under normal conditions it should look like in the first screenshot.

I noticed in the code that the `TripleOscillator` is composed of three `OscillatorObjects` and all of these instances have their own phase offset which means that it is perfectly possible for them to go out of phase, i.e. have different values at certain points in time. I would have assumed that the `TripeOscillator` only keeps one phase accumulator for all its oscillators (at least in mix mode). However, it's also possible that all this is by design and therefore an intended feature.

**Comment by Umcaruje** [comment]
Yeah, this is a bug. The oscillators should stay in phase if you do not change the phase offset manually.

**Comment by grejppi** [comment]
This is likely to be a floating point precision thing. Getting the frequency for a note involves a lot of divisions, multiplications, powf() calls and whatnot. The tiniest errors add up and result in the oscillators getting out of phase.

I made a [test project](https://gist.github.com/grejppi/0bf7818a24c48d76f64d) with two oscillators. The first one has no detuning and plays the notes D4, D4, D4. The second one has its coarse detuning automated as 0, -12, -24 and playing the notes D4, D5, D6.

This is their difference.

![phase difference](https://cloud.githubusercontent.com/assets/772336/8533310/eff4167c-243c-11e5-8c01-210228dd2b83.png)

**Comment by softrabbit** [comment]
Rendering that test project, I get a perfectly silent .wav. Can't see any action on the master output "scope" when playing it either. @grejppi, am I doing something wrong? This on master from approx. yesterday (Linux x86_64, Qt 5.4.1, GCC 4.9.2).

**Comment by michaelgregorius** [comment]
@grejppi The file works for me but I think the problem is not about using two instances of the Triple Osc. If I change the Triple Oscs in your example to only play a sine wave and move the phase of one of the synths by 180° I get perfect silence when playing in LMMS as is expected. In other words: the phases of the two Triple Oscs do not go out of sync (at least in that small example).

I also realized that export seems to be broken because when I render the perfect silence described above I get something that is definitively not silence:
![2047-phase-test](https://cloud.githubusercontent.com/assets/9293269/8551655/3d83273e-24d8-11e5-9776-2bfd6ab7dd8a.png)
So it seems that currently we should not rely on rendered output when analyzing these problems. But I guess that's something for another issue.

If the problem is really caused by floating point precision then the code definitively needs refactoring. Normally you calculate the frequency from the MIDI note value and then initialize a phasor using the frequency and then only increment the phasor using an addition (and a boundary check) with each sample tick. Doing so should work for quite some time without any precision issues.

I also noted that I wrote nonsense in the last section of my comment from 12.05.2015. If the oscillators are tuned differently we obviously need separate phase accumulators for each oscillator. However, if they are tuned in octaves apart they should always sync up once the oscillator with the lowest tuning has finished a cycle.

**Comment by softrabbit** [comment]
I get phase drift when trying the original recipe from @musikBear, an 18 bar D3. And it looks to me like the drift increases with increasing sampling frequency; can't notice anything when exporting 44100x1, but with more oversampling the problem gets stronger. Smells like precision, rounding or something of that kind.

**Comment by softrabbit** [comment]
OK, now I'm sure. It's a precision problem, see https://gist.github.com/softrabbit/d8232cda4bbc3b61cdbc 
Download, compile and run. Change the one ```#if 0``` to ```#if 1```, compile and run again. The last line shows the phases of the oscillators at 8x oversampling at the end of a 40 second note.

To quote the final lines from both runs on my 32-bit system:
```
0.831466       0.07414  0.0766979 (floats)
2.85105e-10    1        1         (doubles, NB. 0==1 when it comes to phase here)
```

Looks like this is one of the places where a 32-bit float isn't quite adequate.

**Comment by musikBear** [comment]
Very nice bughunt there :+1:

**Comment by karmux** [comment]
Which parts of core are affected by this bug? Oscillator class for sure but what else? How much work do you think is to fix it? Would be interesting to hear if fixing this makes sound more punchy for other instruments also.

**Comment by musikBear** [comment]
@karmux good question! But is this fix added to Master? The ticket is still open
I see this
![3ocflux](https://cloud.githubusercontent.com/assets/6662066/20927066/12d378ec-bbc0-11e6-8adf-5d53ed043e17.JPG)
But this is over 72 bar!
I tried twice. Same shape.

**Comment by karmux** [comment]
@musikBear I'm on master and this is what very long D3 note made in TripleOscillator (using settings on screenshot above) gives:

![screenshot_20161206_195929](https://cloud.githubusercontent.com/assets/2575235/20937298/0599871e-bbef-11e6-80a3-341aef94a180.png)
Exported using default settings.

![screenshot_20161206_195907](https://cloud.githubusercontent.com/assets/2575235/20937290/fadf2900-bbee-11e6-93fb-5d96c675f03e.png)
Exported using highest settings.

**Comment by michaelgregorius** [comment]
This should be fixed with PR #3145. Here's the result of a long D3 as described in @musikBear's original report. Top is before the changes with float precision. Bottom is the fixed version where the phase variables have been replaced with doubles:
![2047-comparison-float-vs-double](https://cloud.githubusercontent.com/assets/9293269/20947323/ea7a0c36-bc0e-11e6-9c5b-4407f5c0d028.png)

I have also repeated my test with the three added sine waves and they now also reproduce correctly after longer time spans.

Kudos to @grejppi and @softrabbit for their investigations! :)

**Comment by musikBear** [comment]
@karmux That is almost as weird as mine..
@michaelgregorius Yours looks perfect! If *that* are current Master then @grejppi and @softrabbit has crushed this bug, and the the ticket should be closed.

**Comment by Umcaruje** [comment]
@musikbear no, the screen shot is from a PR that @michaelgregorious
proposed yesterday and it has not yet have been merged.

**Comment by michaelgregorius** [comment]
Please note that using `double` values will not solve the problem completely. I was curious what caused the triangle pattern in @karmux' screenshot at high quality settings and it seems that its caused by the high sample rate which leads to even more phase updates during a given time span. Here you can see a comparison for the test tone with `float` (top) and `double` (bottom) phase updates using a sample rate of 192kHz during the export:
![2047-comparison-f-vs-d-192khz](https://cloud.githubusercontent.com/assets/9293269/20982484/124ae5d2-bcb9-11e6-8a92-d3966f9489b7.png)

I you look at the screenshot at the original size you will see that the bottom waveform also starts to fade. With `float` the phases just go out of phase quicker than with `double` but in the long run it's also happening when using `double` types. However, this seems to be normal. For comparison here's five minutes of the test tone created with the Spire VST and exported at 192kHz:
![2047-spire-at-192khz](https://cloud.githubusercontent.com/assets/9293269/20982551/5b75b318-bcb9-11e6-9e7a-bddb4666df43.png)

Last but not least, playing a single tone for that long does not seem to be that musical anyway. Unless you are an artist in the spirit of John Cage I guess. ;)

**Comment by musikBear** [comment]
@michaelgregorius Again the bottom screenie using double, looks perfect. 
That said, i cant help wondering how different *existing* project using old 3oc will sound, with the repaired 3oc version

**Comment by michaelgregorius** [comment]
For some patches the difference in sound is already quite noticeable for not so long notes. For example if you take the default preset of 3OSC and set all three oscillators to saw then the current `master` will sound phasey and weak very quick. With the patch applied it sounds rock solid even if you hold the note for a long time. I have created a quick demo for this. The first two notes are a rather current `master` without the patch. The next two notes are the same `master` with the patch applied:
https://soundcloud.com/blitbit/techdemo-2047/s-349L7

I have also performed a test with the song "Tectonic" by Farbro from the demo folder as it uses several instances of 3OSC. I have rendered it once with the patch applied and once without. I have then inverted the phase on one of the renders and mixed them down. If there was no difference between the tracks this would yield perfect silence. However, this is the result of the action:
https://soundcloud.com/blitbit/difference-of-tectonic-by-farbro/s-AWQBk

As you can hear the bass seems to be quite affected by the change. To get an idea of how much random elements could influence the result I have also rendered the double version twice and then performed the actions. The result was that only few elements are affected by the randomness. For example the bass did not sound in this case.

So the patch would definitively have an effect on some songs. For example if a song author used the "phaseyness" created by the oscillators as part of his sound design then this aspect of the sound will be gone with the patch.

The following instruments use the class `Oscillator` and might be affected by the changes:

- LB302
- Monstro
- NES (but only for noise it seems)
- Organic
- 3OSC

**Comment by softrabbit** [comment]
> So the patch would definitively have an effect on some songs. For example if a song author used the "phaseyness" created by the oscillators as part of his sound design then this aspect of the sound will be gone with the patch.

OTOH, that phaseyness would've behaved differently for different export settings. So I'd say it's no big loss.

**Comment by musikBear** [comment]
@michaelgregorius super job! 🎆  
I really like how your double fix has made the instrument completely solid in expression, the fact that it will influence existing songs is ofcause a bit problematic. @softrabbit conclusion is correct, but still it is a problem.

**Comment by tresf** [comment]
> the fact that it will influence existing songs is ofcause a bit problematic
> @tresf -perhaps add a paragraph on this in the release-notes.

It's the cost of fixing bugs.  I wouldn't consider it problematic. 

I'm not sure how to explain this in a way that would properly warn people.  The fix will be mentioned, but the end user effect of the fix, that will just have to be observed first hand. :\

**Comment by softrabbit** [comment]
> The following instruments use the class Oscillator and might be affected by the changes:

LB302 and Monstro should be unaffected, they handle the phase by themselves and just look up values from Oscillator. NES is probably okay if it only uses the noise waveform.

That leaves Organic and TripleOscillator, which most certainly are affected.

**Comment by musikBear** [comment]
I have an idea. I will just share it.
Preserve the old oscillator class, and then let lmms select the old class, if the project time-stamp is older, than the 1.2 release-date.
Old projects would then be preserved, unless the user saves it in 1.2
New projects would be with @michaelgregorius 👍 good fix

**Comment by Spekular** [comment]
People who need the old sound can finish their projects in an old version or bounce to audio. I don't think it's worth it to make the code vase messier for something like this.

**Comment by tresf** [comment]
> Preserve the old oscillator class

No.  Sometimes we choose to preserve bugs for backwards compat.  This one is a terrible idea.  It adds unnecessary complexity to our codebase.  As @Spekular said, if people have a problem with this fix, they need to revert to an older version which exhibited the bug.

**Comment by michaelgregorius** [comment]
> LB302 and Monstro should be unaffected, they handle the phase by themselves and just look up values from Oscillator. NES is probably okay if it only uses the noise waveform.

That's correct. However, LB302 and Mostro also store their phase information in `float` types which means that they are still susceptible to the same problems as 3OSC even with the fix applied. If we want to change this I propose that we create new issues for these instruments.

Concerning the backwards compatibility: I also think that keeping both versions of the class would only bloat and complicate the code. Perhaps it makes sense to add some warning to the dialog that pops up when you open an older file with a newer version. Something along the lines of: "You are loading a file created with version X in version Y. Some instruments might sound different." Although I am not sure whether this dialog still exists as I was not able to trigger it with any of the demo songs in a development build.

**Comment by michaelgregorius** [comment]
I have done some more testing with LB302 and Monstro and interestingly they do not exhibit the same problems as 3OSC although they use `float` types for their phase information. So it seems like the problem with 3OSC might in part be caused by the way the phase information is updated and not only due to the usage of `float` instead of `double`.

This also means that there is no need for new issues. :)

**Comment by tresf** [comment]
>  You are loading a file created with version X in version Y. Some instruments might sound different

#432, #2656  ;).

**Comment by tresf** [comment]
Closed via #3145.

**Comment by michaelgregorius** [comment]
Reopening this one as I have reverted the changes from e3e14bb730440fd577b6800c0928f255ac52df63 due to the problems described in #3292.

**Comment by PhysSong** [comment]
I found something was wrong in #3145: `m_ext_phaseOffset` is `const double &`, but some plugins still pass `float` variable to the constructor. If I change them to `double`, that works great and #3292 doesn't occur anymore.

**Comment by michaelgregorius** [comment]
Great find @PhysSong! Please create a pull request so that we can double check for #3292.

**Comment by PhysSong** [comment]
Okay... Target to `master`, or `stable-1.2`? I guess this issue isn't serious.

**Comment by zonkmachine** [comment]
`stable-1.2` I think. This is a regression that many have pointed out.

**Comment by lukas-w** [comment]
Are you sure it is a regression? AFAIK we've always been using `float` for this.

Btw, if we want to increase precision further (as even `double` will have measurable error eventually especially with high sample rates as @michaelgregorius pointed out in https://github.com/LMMS/lmms/issues/2047#issuecomment-265543960), we could make use of a special algorithm such as [Kahan summation](https://en.wikipedia.org/wiki/Kahan_summation_algorithm). This keeps the error down to a constant value (meaning the error is indepent from the number of samples and will not therefore not increase in time), so `float` is probably sufficient if we use that algorithm. Boost provides an implementation in its [Accumulators](http://www.boost.org/doc/libs/1_65_1/doc/html/accumulators.html) library.

**Comment by PhysSong** [comment]
> as even `double` will have measurable error eventually especially with high sample rates

I think it can be fixed by using the method I mentioned above. I will make a quick test for that.

**Comment by PhysSong** [comment]
I've tested with `double` and it showed perfect sync for 1 minute note with 7040Hz(A8), sampling rate 192000Hz, oversampling x8. Theoretically, the error from `double` is negligiby small, even for one-day-long notes, if I remember correctly.

**Comment by zonkmachine** [comment]
> Are you sure it is a regression? AFAIK we've always been using float for this.

No. I took it from memory. Memory wrong.

**Comment by zonkmachine** [comment]
@PhysSong Since it looks like you've basically solved this, maybe bump it to 1.2.0 ?

**Comment by PhysSong** [comment]
Note that fixing this issue may cause performance degrade since it replaces single-precision(`float`) calculations with double-precision(`double`) ones.

**Comment by SecondFlight** [comment]
@PhysSong What's the status on this? I don't see a linked PR for the fix you mentioned, but if the fix is in master or stable 1.2 I'd like to close this issue.

**Comment by PhysSong** [comment]
I haven't added the patch because it may lead to performance degrades due to the change from `float` to `double`.

**Comment by softrabbit** [comment]
Performance probably doesn't suffer much. See https://stackoverflow.com/questions/4584637/double-or-float-which-is-faster for a discussion. My proof of concept code in a comment above, https://github.com/LMMS/lmms/issues/2047#issuecomment-120009714, with slight modifications to run longer and print less gave results like this (seconds of user time):
![Screenshot from 2019-10-12 20-58-53](https://user-images.githubusercontent.com/6411463/66705730-30d6a880-ed33-11e9-8b1f-4539b74e36a7.png)

**Comment by michaelgregorius** [comment]
I think the phase update code is only a very small part of what LMMS is doing when it's playing a song and I am pretty sure that the code with the double implementation will never show up in profiler runs.

To quote Donald Knuth: _"The real problem is that programmers have spent far too much time worrying about efficiency in the wrong places and at the wrong times; premature optimization is the root of all evil (or at least most of it) in programming."_

Concerning potentially weaker hardware like ARM: If a user wants to run LMMS on a weaker hardware then that user should pay the price for that and not all of LMMS' users.

**Comment by DeRobyJ** [comment]
I we really don't want to switch to double, we could at least work around the LFO syncing problem by secretly _faking_ it.

Like resetting the value to initial phase every time we get to a new measure.
This behavior would be activated once the Sync to Tempo feature is used, and it would be deactivated as soon as the LFO speed is changed in any other way.

A more proper, but quite longer, fix would be to refactor this feature to calculate values based on position rather than adding it up over time...

**Comment by michaelgregorius** [comment]
Hi @PhysSong,

can you please create a pull request of your changes so that we can put them through a profiler?

To be frank, I find it a bit sad that this issue is five years old now and that it's still open although we know what the problem is, what the solution is and are very close to fixing it. I am pretty sure that if the oscillator had been implemented from the start using `double` precision no one would have gotten the idea to replace the `double` with `float` for performance reasons, especially if it worsens the quality of the oscillator by that much.

**Comment by zonkmachine** [comment]
> __
> 
> Concerning potentially weaker hardware like ARM: If a user wants to run LMMS on a weaker hardware then that user should pay the price for that and not all of LMMS' users.

I'm one of those users on a weaker machine. I vote double precision.

**Comment by PhysSong** [comment]
I started working on this again. Should I replace all frequency calculations involving `float` to `double` to guarantee precise phase for other instruments as well?

**Comment by michaelgregorius** [comment]
Hi @PhysSong, I think this issue can be closed. I have repeated the original test described at the start of this issue and got a solid signal.

**Comment by zonkmachine** [comment]
stable-1.2 vs.
master d71116b82
![sawdetune](https://github.com/LMMS/lmms/assets/6368949/d49ccc02-a79d-49e5-bc56-e85b21cbea60)

The issue has improved greatly but it's still showing a hint of the issue mentioned. It's clearly improved so I'm pushing this to 1.3+.

Demo project after @musikBear 's instructions: [sawdetune.mmp.txt](https://github.com/user-attachments/files/15752551/sawdetune.mmp.txt)

**Comment by musikBear** [comment]
> The issue has improved greatly but it's still showing a hint of the issue mentioned. 

Just to make sure that i understand the image:
The top part is _before_ the improvement, and the bottom-part _is with_ the improvements?
If so it is almost fixed. 👍

**Comment by zonkmachine** [comment]
> The top part is _before_ the improvement, and the bottom-part _is with_ the improvements?

Yes. Top track is on stable-1.2 and the bottom one on master (https://github.com/LMMS/lmms/commit/d71116b82a0a20355eb0978b8fb5379a368ba2af).
